docs/design/001-prometheus.md

```diff
@@ -0,0 +1,451 @@
+- name: executed_processes
+  type: counter
+  category: trace
+  gadget: exec
+  labels:
+    - k8s.pod
+    - k8s.container
+  selector:
+    - "k8s.namespace:default"
+```
+
+Or only count events for a given command:
+
+```yaml
+# cat executions by namespace, pod and container
+- name: executed_cats # ohno!
+  type: counter
+  category: trace
+  gadget: exec
+  labels:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+  selector:
+    - "comm:cat"
+```
+
+And finally, we can provide counters for failed operations:
+
+```yaml
+# failed execs by namespace, pod and container
+- name: failed_execs
+  type: counter
+  category: trace
+  gadget: exec
+  labels:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+  selector:
+    - "retval:!0"
+```
+
+Filtering can also be used for gadgets that provide events describing two different situations, for
+instance the trace dns gadget emits events for requests and answers. Then, we can expose a counter
+only for requests based on the value of the "qr" field.
+
+```yaml
+# DNS requests aggregated by namespace and pod
+- name: dns_requests
+  type: counter
+  category: trace
+  gadget: dns
+  labels:
+    - namespace
+    - pod
+  selector:
+    # Only count query events
+    - "qr:Q"
+```
+
+Another example is:
+
+```yaml
+  # bpf seccomp violations
+- name: seccomp_violations
+  type: counter
+  category: audit
+  gadget: seccomp
+  labels:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+    - syscall
+  selector:
+    - "syscall:bpf"
+```
+
+By default, a counter is increased by one each time there is an event, however it's possible to
+increase a counter using a field on the event:
+
+```yaml
+# Read bytes on ext4 filesystem
+- name: read_bytes_ext4
+  type: counter
+  category: trace
+  gadget: fsslower
+  labels:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+  field: bytes
+  selector:
+    - "filesystem:ext4"
+    - "op:R"
+```
+
+## Gauges
+
+"A _gauge_ is a metric that represents a single numerical value that can arbitrarily go up and down"
+from
+[https://prometheus.io/docs/concepts/metric_types/#gauge](https://prometheus.io/docs/concepts/metric_types/#gauge).
+
+It seems that the only category of gadgets that can provide data to be interpreted as a gauge is the
+snapshotters.
+
+```yaml
+# Number of processes by namespace / pod / container
+- name: number_of_processes
+  type: gauge
+  category: snapshot
+  gadget: process
+  aggregate:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+
+# Number of sockets in CLOSE_WAIT state
+- name: number_of_sockets_close_wait
+  type: gauge
+  category: snapshot
+  gadget: socket
+  aggregate:
+    - k8s.namespace
+    - k8s.pod
+    - k8s.container
+  selector:
+    - "status:CLOSE_WAIT"
+```
+
+TODO: This is not totally clear how this should work since there gadget doesn't provide a stream of
+events. In this case we should execute the gadget each time prometheus scrapes the endpoint.
```

Comment:
I guess they could be run like the top category in that case - using an interval. I wouldn't do things "on scrape", because of the delay this could cause (and thus skew the other metrics).

Reply:
This is something we should think more about, specially  regarding the "cache" to avoid long times on scrapping. I remember I read somewhere clients shouldn't implement that kind of logic, but I can't find it again.